Power iteration: check norm of initial vector

peterrum opened the pull request:
... and early return if norm is zero. This might happen if all degrees of freedoms (on a coarse grid/single cell) are constrained.

peterrum committed:
Power iteration: check norm of initial vector

bangerth commented on include/deal.II/lac/precondition.h:
```diff
@@ -2397,7 +2397,14 @@ namespace internal
   {
     typename VectorType::value_type eigenvalue_estimate = 0.;
-    eigenvector /= eigenvector.l2_norm();
+
+    const auto norm = eigenvector.l2_norm();
+
+    if (norm == 0.0)
+      return 1.0;
+
+    eigenvector /= norm;
```
I think that's the wrong solution. If you pass in a zero vector by accident, the function pretends that the largest eigenvalue of the operator is one. That's just not true.

The function should probably document that one is must provide a non-zero vector and that if the vector *is* zero throw an exception that one can catch.

peterrum commented:
> I think that's the wrong solution. If you pass in a zero vector by accident, the function pretends that the largest eigenvalue of the operator is one. That's just not true.
The function should probably document that one must provide a non-zero vector and that if the vector is zero throw an exception that one can catch.

One should note that this is the internal eigenvalue estimation of `PreconditionChebyshev` and `PreconditionRelaxation`. They use either `EigenvalueAlgorithm::lanczos` or `EigenvalueAlgorithm::power_iteration`. The first one returned for the described system (all dofs are constrained; as sometimes is the case on the coarse grid) `ev_min=ev_max=1.0` (which makes sense if one considers the matrix corresponding to a fully constrained system to be a identity matrix). The second crashed because there is a division by 0. This PR tries to fix this behavior.

In the new commit, I have moved the if-statement from `internal::power_iteration()` to `internal::estimate_eigenvalues()`. Furthermore, I have added a simplified test which checks the desired behavior. 

This is the best solution I can come up. I am open for suggestions how to alternatively fix this issue.

kronbichler reviewed (approved):
I think this is much clearer, so I'd go with this setup.

blaisb reviewed (approved):
Like this a lot. Good work.

drwells commented on include/deal.II/lac/precondition.h:
```diff
@@ -2451,44 +2451,52 @@ namespace internal
-        if (data.eigenvalue_algorithm == internal::EigenvalueAlgorithm::lanczos)
+        // note: temp_vector1 only contains zeroes, which happens if all dofs
+        // are constrained, there is nothing to do
+        if (temp_vector1.l2_norm() != 0.0)
```
Is it possible to use `all_zero()` here? We do require that for the vector space vector concept (which this template doesn't check for).

bangerth replied:
Indeed, that should be cheaper.

bangerth commented on include/deal.II/lac/precondition.h:
```diff
@@ -2451,44 +2451,52 @@ namespace internal
         data.constraints.set_zero(temp_vector1);
 
-        if (data.eigenvalue_algorithm == internal::EigenvalueAlgorithm::lanczos)
+        // note: temp_vector1 only contains zeroes, which happens if all dofs
+        // are constrained, there is nothing to do
+        if (temp_vector1.l2_norm() != 0.0)
```
I think the comment has the wrong grammar. "`temp_vector1` only contains zeroes" is a statement, but the statement is not generally true. I think what you want to say is something like this:
```suggestion
        // We want to run the power iteration starting with temp_vector1, but for that it
        // needs to be a nonzero vector. However, there are situations where it only contains
        // zeroes (e.g., if all dofs are constrained). If that is the case, there is nothing to do.
        if (temp_vector1.l2_norm() != 0.0)
```

bangerth reviewed (commented):
Yes, this is much clearer.

bangerth commented:
Do you think we could add a test that checks that this continues to work?

bangerth commented:
@peterrum Ping?

bangerth commented:
@peterrum Ping?

peterrum commented:
I have updated the comment.

> Do you think we could add a test that checks that this continues to work?

What do you mean? I have added a new test.

bangerth commented on include/deal.II/lac/precondition.h:
```diff
@@ -2454,44 +2454,54 @@ namespace internal
 
-        if (data.eigenvalue_algorithm == internal::EigenvalueAlgorithm::lanczos)
+        /// We want to run the power iteration starting with temp_vector1, but
+        /// for that it needs to be a nonzero vector. However, there are
```
Replace `///` by `//`.

bangerth commented on tests/lac/precondition_chebyshev_08.cc:
```diff
@@ -0,0 +1,64 @@
+  const auto     ev = precon.estimate_eigenvalues(vec);
+
+  AssertDimension(ev.min_eigenvalue_estimate, 1.0);
+  AssertDimension(ev.max_eigenvalue_estimate, 1.0);
```
I don't think `AssertDimension` is meant to be used for comparing floating point equality. Let's just explicitly compare to 1.0 -- it's a floating point comparison, but we set these numbers explicitly to 1.0, so checking for equality should be ok.

bangerth commented on include/deal.II/lac/precondition.h:
```diff
@@ -2454,44 +2454,54 @@ namespace internal
+        /// We want to run the power iteration starting with temp_vector1, but
+        /// for that it needs to be a nonzero vector. However, there are
+        // situations where it only contains zeroes (e.g., if all dofs are
+        // constrained). If that is the case, there is nothing to do.
```
This sentence sent me down a ten-minute rabbit hole.  I think what you're trying to say is "if we cannot initialize the random starting vector to anything nonzero, then there is nothing *we can do*", rather than "there is nothing that needs to be done". Perhaps this?

```suggestion
        // constrained). If that is the case, there is nothing we can do because in both of the
        // two methods below, we either solve a linear system with temp_vector1 as the right
        // hand side, or use it as a starting guess for a power iteration, but in neither case a
        // zero vector yields anything useful. So skip the following code and instead use the
        // fallback below if temp_vector1 is the zero vector.
```

bangerth reviewed (commented):
@peterrum I looked at this again. The code makes sense to me, but I got massively hung up on one comment. Does the suggested change to the comment make sense to you? If you're already there, would you mind also addressing the other comments? I don't think this should take you more than 5 minutes in total, and then we can merge this PR.

Ping me when you've made the changes!

kronbichler commented:
This looks almost ready for inclusion in the release. @peterrum do you think we can finalize this one?

peterrum commented:
@bangerth @kronbichler Done! Thx!
